### DISCUSSION TITLE
Naming convention for constraints in ORM

### DISCUSSION THREAD

**Discussion by ziima** [sqlalchemy/sqlalchemy#10714] [Usage Questions]
I have troubles to set up a reasonable naming convention on declarative ORM classes which constitutes both boolean constraints and custom constraints.

Here's a MWE:
```python
from sqlalchemy import Boolean, CheckConstraint, MetaData
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
from sqlalchemy import create_engine

constraint_naming_conventions = {
    "ix": "ix_%(column_0_label)s",
    "uq": "uq_%(table_name)s_%(column_0_name)s",
    "ck": "ck_%(table_name)s_%(constraint_name)s",
    # "ck": "ck_%(table_name)s_%(column_0_name)s_%(constraint_name)s",
    "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
    "pk": "pk_%(table_name)s",
}


class Base(DeclarativeBase):
    metadata = MetaData(naming_convention=constraint_naming_conventions)

    type_annotation_map = {
        bool: Boolean(create_constraint=True, name='bool'),
    }

    id: Mapped[int] = mapped_column(primary_key=True)


class ModelAlpha(Base):
    __tablename__ = "alpha"

    is_true: Mapped[bool]
    is_false: Mapped[bool]
    value: Mapped[int] = mapped_column(CheckConstraint('value > 5', name='min'))


engine = create_engine("sqlite:///:memory:", echo=True)
Base.metadata.create_all(bind=engine)
```

Based on docs https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html#creating-indexes-and-constraints-with-naming-conventions-on-mixins, I started with `"ck_%(table_name)s_%(constraint_name)s"` which works fine for `CheckConstraint`s, but it generates the same constraint name on boolean constraints in the single table. I tried to update the convention to `"ck_%(table_name)s_%(column_0_name)s_%(constraint_name)s"` based on https://docs.sqlalchemy.org/en/20/core/constraints.html#configuring-naming-for-boolean-enum-and-other-schema-types, but found it doesn't work on `CheckConstraint`, since it's not bound to any column.

SQLs generated with naming convention: `"ck_%(table_name)s_%(constraint_name)s"` (two constraints with the same name):
```
CREATE TABLE alpha (
	is_true BOOLEAN NOT NULL, 
	is_false BOOLEAN NOT NULL, 
	value INTEGER NOT NULL CONSTRAINT ck_alpha_min CHECK (value > 5), 
	id INTEGER NOT NULL, 
	CONSTRAINT pk_alpha PRIMARY KEY (id), 
	CONSTRAINT ck_alpha_bool CHECK (is_true IN (0, 1)), 
	CONSTRAINT ck_alpha_bool CHECK (is_false IN (0, 1))
)
```
with naming convention: `"ck_%(table_name)s_%(column_0_name)s_%(constraint_name)s"` (missing column name in custom constraint)
```
CREATE TABLE alpha (
	is_true BOOLEAN NOT NULL, 
	is_false BOOLEAN NOT NULL, 
	value INTEGER NOT NULL CONSTRAINT ck_alpha__min CHECK (value > 5), 
	id INTEGER NOT NULL, 
	CONSTRAINT pk_alpha PRIMARY KEY (id), 
	CONSTRAINT ck_alpha_is_true_bool CHECK (is_true IN (0, 1)), 
	CONSTRAINT ck_alpha_is_false_bool CHECK (is_false IN (0, 1))
)
```
I'd like
```
CREATE TABLE alpha (
	is_true BOOLEAN NOT NULL, 
	is_false BOOLEAN NOT NULL, 
	value INTEGER NOT NULL CONSTRAINT ck_alpha_value_min CHECK (value > 5), 
	id INTEGER NOT NULL, 
	CONSTRAINT pk_alpha PRIMARY KEY (id), 
	CONSTRAINT ck_alpha_is_true_bool CHECK (is_true IN (0, 1)), 
	CONSTRAINT ck_alpha_is_false_bool CHECK (is_false IN (0, 1))
)
```

**Comment by CaselIT** [comment]
Hi,

You can use a custom function if you prefer, or use `column('value') > 5` instead of a generic text in the contraint definition so that a column name be obtained

**Reply by ziima** [reply]
Custom function seems a bit like overkill for a single constraint.

I tried `CheckConstraint(column('value') > 5, name='min')`, but I end up in an `AssertionError`:
```
Traceback (most recent call last):
  File "/tmp/contraints_naming.py", line 25, in <module>
    class ModelAlpha(Base):
  File "/tmp/contraints_naming.py", line 30, in ModelAlpha
    value: Mapped[int] = mapped_column(CheckConstraint(column('value') > 5, name='min'))
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vzima/.virtualenvs/fred/lib/python3.11/site-packages/sqlalchemy/orm/_orm_constructors.py", line 314, in mapped_column
    return MappedColumn(
           ^^^^^^^^^^^^^
  File "/home/vzima/.virtualenvs/fred/lib/python3.11/site-packages/sqlalchemy/orm/properties.py", line 585, in __init__
    self.column = cast("Column[_T]", Column(*arg, **kw))
                                     ^^^^^^^^^^^^^^^^^^
  File "/home/vzima/.virtualenvs/fred/lib/python3.11/site-packages/sqlalchemy/sql/schema.py", line 2150, in __init__
    self._init_items(*cast(_typing_Sequence[SchemaItem], l_args))
  File "/home/vzima/.virtualenvs/fred/lib/python3.11/site-packages/sqlalchemy/sql/schema.py", line 231, in _init_items
    spwd(self, **kw)
  File "/home/vzima/.virtualenvs/fred/lib/python3.11/site-packages/sqlalchemy/sql/base.py", line 1337, in _set_parent_with_dispatch
    self._set_parent(parent, **kw)
  File "/home/vzima/.virtualenvs/fred/lib/python3.11/site-packages/sqlalchemy/sql/schema.py", line 4370, in _set_parent
    ColumnCollectionMixin._set_parent(self, parent)
  File "/home/vzima/.virtualenvs/fred/lib/python3.11/site-packages/sqlalchemy/sql/schema.py", line 4309, in _set_parent
    for col in self._col_expressions(parent):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vzima/.virtualenvs/fred/lib/python3.11/site-packages/sqlalchemy/sql/schema.py", line 4291, in _col_expressions
    assert len(result) == len(self._pending_colargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

**Reply by CaselIT** [reply]
I guess sqlalchemy does not understand that `column('value')` is the same column as the table it's trying to define

@zzzeek do you have suggestions?

**Reply by zzzeek** [reply]
first suggestion is dont use the check constraint feature for Boolean as it's overkill, failing that, give the Boolean type a name, it accepts a name argument for this purpose https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Boolean.params.name

**Reply by zzzeek** [reply]
yah this wont work, I'd rather just remove the constraint feature from booleans, sorry

```
type_annotation_map = {
        bool: Boolean(create_constraint=True, name='bool'),
    }
```

**Reply by zzzeek** [reply]
there maybe in the implementation a codepath here where MappedColumn can pass along the name, in the same way it passes things along to Enum.  I can accept PRs for this feature that have tests , if we can determine that the architecture is straightforward

**Reply by zzzeek** [reply]
it would pass column name here:

https://github.com/sqlalchemy/sqlalchemy/blob/82690b1cfc1e76e5deb622a9afefbcf3be299962/lib/sqlalchemy/orm/properties.py#L859

from there the name would be passed through where it can be optionally used by types in _resolve_for_python_type where a differentiating name is helpful:  

https://github.com/sqlalchemy/sqlalchemy/blob/82690b1cfc1e76e5deb622a9afefbcf3be299962/lib/sqlalchemy/sql/sqltypes.py#L1525

however we'd have to figure out what to do with Enum vs. Boolean, could get weird

**Reply by ziima** [reply]
> first suggestion is dont use the check constraint feature for Boolean as it's overkill

So the usage of `create_constraint` in Boolean is discouraged? I would at least made a PR to note that in the docs.

**Reply by zzzeek** [reply]
it's not "discouraged" but we did change the default behavior for this flag from True to False, so that people didn't inherit these constraint issues automatically (particularly on the Alembic side).    I would say the feature is there if a team wants to try to enforce their boolean cols to be constrained, but that this represents a little more formality than everyone might want.
